### ISSUE TITLE
Cannot create an instance of the abstract class 'FriendRequest

### ISSUE THREAD

**Issue by JasLin** [wechaty/wechaty#70]
i am trying to test  FriendRequest in ``` example/ding-dong-bot.js`` , error occurs like that:

```
example/ding-dong-bot.ts (49,19): Cannot create an instance of the abstract class 'FriendRequest'. (2511)
```

code: 
```example/ding-dong-bot.js``` : 

```javascript
/**
 *
 * Wechaty - Wechat for Bot
 *
 * Connecting ChatBots
 * https://github.com/wechaty/wechaty
 *
 */

/* tslint:disable:variable-name */
const QrcodeTerminal = require('qrcode-terminal')

import {
  Wechaty
  , Config
  , log
  , FriendRequest
} from '../'

const welcome = `
| __        __        _           _
| \\ \\      / /__  ___| |__   __ _| |_ _   _
|  \\ \\ /\\ / / _ \\/ __| '_ \\ / _\` | __| | | |
|   \\ V  V /  __/ (__| | | | (_| | |_| |_| |
|    \\_/\\_/ \\___|\\___|_| |_|\\__,_|\\__|\\__, |
|                                     |___/

=============== Powered by Wechaty ===============
-------- https://github.com/zixia/wechaty --------

I'm a bot, my super power is talk in Wechat.

If you send me a 'ding', I will reply you a 'dong'!
__________________________________________________

Hope you like it, and you are very welcome to
upgrade me for more super powers!

Please wait... I'm trying to login in...

`

console.log(welcome)
const bot = Wechaty.instance({ profile: Config.DEFAULT_PROFILE })

bot
.on('login'	  , user => {
  log.info('Bot', `${user.name()} logined`)
  const request = new FriendRequest()
  console.log('loging....................................')
  console.log(request)
})
.on('logout'	, user => log.info('Bot', `${user.name()} logouted`))
.on('error'   , e => log.info('Bot', 'error: %s', e))
.on('scan', (url, code) => {
  if (!/201|200/.test(String(code))) {
    let loginUrl = url.replace(/\/qrcode\//, '/l/')
    QrcodeTerminal.generate(loginUrl)
  }
  console.log(`${url}\n[${code}] Scan QR Code in above url to login: `)
})
.on('message', m => {
  try {
    const room = m.room()
    console.log((room ? '[' + room.topic() + ']' : '')
                + '<' + m.from().name() + '>'
                + ':' + m.toStringDigest()
    )

    if (/^(ding|ping|bing)$/i.test(m.content()) && !m.self()) {
      m.say('dong')
      log.info('Bot', 'REPLY: dong')
    }
  } catch (e) {
    log.error('Bot', 'on(message) exception: %s' , e)
  }
})

bot.init()
.catch(e => {
  log.error('Bot', 'init() fail: %s', e)
  bot.quit()
  process.exit(-1)
})

function logToFile(data) {
  require('fs').appendFile('message.log', data + '\n\n#############################\n\n', err => {
    if (err) { log.error('LogToFile: %s', err) }
  })
}
if (typeof logToFile === 'fasdfsd') {
  console.log('disable linting warning')
}
```

**Comment by zixia** [comment]
hmmm... maybe it's a design mistake made by me.

say briefly, you need to use `PuppetWebFriendRequest` instead of `FriendRequest`, like this:

```typescript
import PuppetWebFriendRequest as FriendRequest from '../src/puppet-web/friend-request'
```

However, I agree it seems should use FriendRequest because PuppetWeb interface should not be used by user. I will think about that later.

**Comment by JasLin** [comment]
it seems puppet-web instance was not set to Config collectlly, while calling``` Config.puppetInstance()```.

```
ERR PuppetWebEvent onServerLogin() exception: Error: no puppet instance                                                                   
Error: no puppet instance                                                                                                                 
    at Object.puppetInstance (/home/jas/projects/learning/wechat/wechaty-ts/src/config.ts:147:11)                                         
    at PuppetWebFriendRequest.FriendRequest (/home/jas/projects/learning/wechat/wechaty-ts/src/friend-request.ts:23:17)                   
    at PuppetWebFriendRequest (/home/jas/projects/learning/wechat/wechaty-ts/src/puppet-web/friend-request.ts:38:5)                       
    at Wechaty.bot.on.user (/home/jas/projects/learning/wechat/wechaty-ts/example/ding-dong-bot.ts:51:19)                                 
    at emitOne (events.js:96:13)                                                                                                          
    at Wechaty.emit (events.js:188:7)                                                                                                     
    at PuppetWeb.puppet.on (/home/jas/projects/learning/wechat/wechaty-ts/src/wechaty.ts:243:19)                                          
    at emitOne (events.js:96:13)                                                                                                          
    at PuppetWeb.emit (events.js:188:7)                                                                                                   
    at PuppetWeb.<anonymous> (/home/jas/projects/learning/wechat/wechaty-ts/src/puppet-web/event.ts:312:10) 
```

code: 
```example/ding-dong-bot.js```
```javascript
/**
 *
 * Wechaty - Wechat for Bot
 *
 * Connecting ChatBots
 * https://github.com/wechaty/wechaty
 *
 */

/* tslint:disable:variable-name */
const QrcodeTerminal = require('qrcode-terminal')

import {
  Wechaty
  , Config
  , log
  // , FriendRequest
} from '../'

import PuppetWebFriendRequest  from '../src/puppet-web/friend-request'

const welcome = `
| __        __        _           _
| \\ \\      / /__  ___| |__   __ _| |_ _   _
|  \\ \\ /\\ / / _ \\/ __| '_ \\ / _\` | __| | | |
|   \\ V  V /  __/ (__| | | | (_| | |_| |_| |
|    \\_/\\_/ \\___|\\___|_| |_|\\__,_|\\__|\\__, |
|                                     |___/

=============== Powered by Wechaty ===============
-------- https://github.com/zixia/wechaty --------

I'm a bot, my super power is talk in Wechat.

If you send me a 'ding', I will reply you a 'dong'!
__________________________________________________

Hope you like it, and you are very welcome to
upgrade me for more super powers!

Please wait... I'm trying to login in...

`

console.log(welcome)
const bot = Wechaty.instance({ profile: Config.DEFAULT_PROFILE })

bot
.on('login'	  , user => {
  log.info('Bot', `${user.name()} logined`)
  const request = new PuppetWebFriendRequest()
  console.log('loging....................................')
  console.log(request)
})
.on('logout'	, user => log.info('Bot', `${user.name()} logouted`))
.on('error'   , e => log.info('Bot', 'error: %s', e))
.on('scan', (url, code) => {
  if (!/201|200/.test(String(code))) {
    let loginUrl = url.replace(/\/qrcode\//, '/l/')
    QrcodeTerminal.generate(loginUrl)
  }
  console.log(`${url}\n[${code}] Scan QR Code in above url to login: `)
})
.on('message', m => {
  try {
    const room = m.room()
    console.log((room ? '[' + room.topic() + ']' : '')
                + '<' + m.from().name() + '>'
                + ':' + m.toStringDigest()
    )

    if (/^(ding|ping|bing)$/i.test(m.content()) && !m.self()) {
      m.say('dong')
      log.info('Bot', 'REPLY: dong')
    }
  } catch (e) {
    log.error('Bot', 'on(message) exception: %s' , e)
  }
})

bot.init()
.catch(e => {
  log.error('Bot', 'init() fail: %s', e)
  bot.quit()
  process.exit(-1)
})

function logToFile(data) {
  require('fs').appendFile('message.log', data + '\n\n#############################\n\n', err => {
    if (err) { log.error('LogToFile: %s', err) }
  })
}
if (typeof logToFile === 'fasdfsd') {
  console.log('disable linting warning')
}
```

**Comment by JasLin** [comment]
@zixia maybe we can use a RequestFactory class to create requst instance according to puppet type，since user who using wechaty dose not care about the puppet type?

**Comment by zixia** [comment]
@JasLin good catch! It's a weird bug.

I will continue debugging it tomorrow, with the `FriendRequest` problem.

**Comment by zixia** [comment]
@JasLin do you think we could use any `Dependency Injection` way to solve this?

**Comment by JasLin** [comment]
@zixia   DI may be the most elegant way,but it's  little heavy for this case ? DI is more suitable for some uncertain business, and create instance according to the business with some configuration. in our case, there is only a limited object types (such as friend-request,bridge..)  to be create according to puppet type . and the puppet  is only in several specific types (such as web,android,ios,wechat4u ..?).

**Comment by JasLin** [comment]
but DI will make the code more elegant ,and bring some other benefits.  there is no doubt.

**Comment by JasLin** [comment]
@zixia had fixed this issue, and opened a [pull request](https://github.com/wechaty/wechaty/pull/76) to the project. 

all the test passed in my computer , but the pull request fail fail in CI, cause with some other codes in the base branch.

**Comment by zixia** [comment]
@JasLin thanks for the contribute!

I had said It is a weird bug because:

1. I can reproduce it with the code you provided, 
2. but if I put ` const request = new PuppetWebFriendRequest()` in other places, like `on('message',...)` or `init().then(...)`, It will not throw any exception.

So it is weird because the `_puppetInstnace` should be inited before `Wechaty.init()` done, so it should be ready for any events like `login`.

I worked on it yesterday but found nothing at all.

After I saw your pull request today, which you save `_puppetInstance` to another place, `global[]`. In theory, I believe this should change nothing because the `Config._puppetInstance` is no different with `global[]`. 

Then I go to re-test this bug with the latest code base, you know what? This bug had gone. It becomes weirder than before.

Could you test it again with the latest master?

P.S. you need to change `import` to: `import { FriendRequest as PuppetWebFriendRequest }  from '../src/puppet-web/friend-request'`

**Comment by JasLin** [comment]
save _instance in local and global make the resulet difference:) I had replied u on issue #71

**Comment by JasLin** [comment]
ok,comming...

**Comment by zixia** [comment]
Hi @JasLin ,

I really can not get this exception again with the latest code of master branch. 

To collaborate with you, I had set up a public cloud9 ide workspace, and put the latest wechaty code in it, with your code put in `example/jaslin.ts`.

Ran `npm run ts-node example/jaslin.ts` will test your code. When I ran, there's no exception(weirder).

Could you please go to ide9(the invitation letter should already in your Gmail inbox), and have a look at your code over there?

P.S. the public cloud9 ide workspace for collaboration, is at: 
https://ide.c9.io/zixia/wechaty-collaboration

**Comment by JasLin** [comment]
```
loging....................................
ERR PuppetWebEvent onServerMessage() exception: Error: no puppet instance
    at Object.puppetInstance (/home/ubuntu/workspace/src/config.ts:142:13)
    at PuppetWebFriendRequest.FriendRequest (/home/ubuntu/workspace/src/friend-request.ts:28:17)
    at PuppetWebFriendRequest (/home/ubuntu/workspace/src/puppet-web/friend-request.ts:38:5)
    at Wechaty.bot.on.on.on.on.on.m (/home/ubuntu/workspace/example/jaslin.ts:63:19)
    at emitOne (events.js:96:13)
    at Wechaty.emit (events.js:188:7)
    at PuppetWeb.puppet.on (/home/ubuntu/workspace/src/wechaty.ts:239:19)
    at emitOne (events.js:96:13)
    at PuppetWeb.emit (events.js:188:7)
    at PuppetWeb.<anonymous> (/home/ubuntu/workspace/src/puppet-web/event.ts:384:10)
(node:14515) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: no puppet instance
(node:14515) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by JasLin** [comment]
error form c9 with my code. 

i forgot to backup your code , can you paste it here , i will diff them.

**Comment by zixia** [comment]
ah, PR closes this by mistake...

after collaborating with @JasLin on Cloud9 IDE, finally, we had caught the bug together, haha!

TL;DR
----------------
this problem will occur if match the following two conditions:

1. you use `ts-node` to run wechaty inside wechaty source directory, and
1. you had a `dist` directory with compiled javascript source code. (if you run `npm run build`)

then the module system will load `src/config.ts` module twice, then we will have two instances of `Config`, and the `_puppetInstance` will be unpredictable, because we do not know which one we are using.

Details
---------------------

### 1. let's put those two lines in the beginning of `src/config.ts`

```ts
console.log('\nconfig.ts imported!!!!!!!!!!!!!!' + __dirname + '\n')
console.log(new Error().stack )
```

this code will log when `Config` module is imported.

### 2. run `npm run build` to get a `dist` directory full of javascript

```shell
$ npm run build
```

### 3. run our code by ts-node

```shell
$ npm run ts-node example/zixia.ts

> wechaty@0.5.1 ts-node /home/ubuntu/workspace
> ts-node "example/zixia.ts"


config.ts imported!!!!!!!!!!!!!!/home/ubuntu/workspace/dist/src

Error
    at Object.<anonymous> (/home/ubuntu/workspace/src/config.ts:8:13)
    at Module._compile (module.js:573:32)
    at Object.Module._extensions..js (module.js:582:10)
    at Module.load (module.js:490:32)
    at tryModuleLoad (module.js:449:12)
    at Function.Module._load (module.js:441:3)
    at Module.require (module.js:500:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/ubuntu/workspace/index.ts:1:1)
    at Module._compile (module.js:573:32)

config.ts imported!!!!!!!!!!!!!!/home/ubuntu/workspace/src

Error
    at Object.<anonymous> (/home/ubuntu/workspace/src/config.ts:8:13)
    at Module._compile (module.js:573:32)
    at Module.m._compile (/home/ubuntu/workspace/node_modules/ts-node/src/index.ts:406:23)
    at Module._extensions..js (module.js:582:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/home/ubuntu/workspace/node_modules/ts-node/src/index.ts:409:12)
    at Module.load (module.js:490:32)
    at tryModuleLoad (module.js:449:12)
    at Function.Module._load (module.js:441:3)
    at Module.require (module.js:500:17)
    at require (internal/module.js:20:19)

## Solution

Will dig it deeper later. now the best way to avoid this is:

1. if use `ts-node`, then `rm -fr dist` first
1. or run `node example/zixia.js` inside `dist` directory.

Thank you @JasLin, Cheers!

**Comment by JasLin** [comment]
@zixia  ur welcome.

so let the instance save to global until u fixed it?

**Comment by zixia** [comment]
@JasLin no problem, let's keep this bug compatible for now. :)

**Comment by zixia** [comment]
@JasLin 

another fix: change your import code to the following, will keep the `Config` instance to one:

```ts
import {
    Config
  , log
} from '../src/config'

import {   Wechaty } from '../src/wechaty'
import { FriendRequest as PuppetWebFriendRequest } from '../src/puppet-web/friend-request'
```

the key difference is:
```diff
import {
    Config
  , log
- } from '../'
+ } from '../src/config'
```

The reason is that if you are `import ... from '../'` and you have a `dist` directory, then your code will load JavaScript module. However, your `import { FriendRequest } from '../src/puppet-web/friend-request'` will load the TypeScript module.

At last, you will have two instances of `Config`: one is JavaScript version, another is TypeScript version.

This bug will not appear if you are not run your program inside wechaty source directory. Because if you are using `import { Wechaty, FriendRequest } from 'wechaty'`, you can only get one instance of `Config` then.

**Comment by JasLin** [comment]
great！i will test this tomorrow,and make the feedback:)

**Comment by JasLin** [comment]
@zixia confirm it, this issue has gone.

**Comment by JasLin** [comment]
@zixia  maybe we should import ```wechaty/src/puppet-web/friend-request``` as ```friend-request``` in ```index.ts``` at this time for user, before found some other solution?

index.ts
```javascript
import { FriendRequest }  from 'wechaty/src/puppet-web/friend-request
```

i make an facade.ts in my source root directory, to fix the path reference problem while  run wechaty with docker or with source at the same time , 

facade.docker.ts
```javascript
import {
    Config
  , Sayable
  , log
}                         from 'wechaty/src/config'
import { Contact }        from 'wechaty/src/contact'
import { FriendRequest }  from 'wechaty/src/puppet-web/friend-request'
import { IoClient }       from 'wechaty/src/io-client'
import { Message }        from 'wechaty/src/message'
import { Puppet }         from 'wechaty/src/puppet'
import { PuppetWeb }      from 'wechaty/src/puppet-web/'
import { Room }           from 'wechaty/src/room'
import { UtilLib }        from 'wechaty/src/util-lib'
import { Wechaty }        from 'wechaty/src/wechaty'
import { StateMonitor }   from 'wechaty/src/state-monitor'
import { Brolog}          from 'wechaty/src/brolog-env'

export {
    Config
  , Contact
  , FriendRequest
  , IoClient
  , Message
  , Puppet
  , PuppetWeb
  , Room
  , Sayable
  , UtilLib
  , Wechaty
  , log // for convenionce use npmlog with environment variable LEVEL
  , Io
  , StateMonitor
}
````

facade.source.ts
```javascript 
import {
    Config
  , Sayable
  , log
}                         from '../src/config'
import { Contact }        from '../src/contact'
import { FriendRequest }  from '../src/puppet-web/friend-request'
import { IoClient }       from '../src/io-client'
import { Message }        from '../src/message'
import { Puppet }         from '../src/puppet'
import { PuppetWeb }      from '../src/puppet-web/'
import { Room }           from '../src/room'
import { UtilLib }        from '../src/util-lib'
import { Wechaty }        from '../src/wechaty'
import { StateMonitor }   from '../src/state-monitor'
import { Brolog}          from '../src/brolog-env'

export {
    Config
  , Contact
  , FriendRequest
  , IoClient
  , Message
  , Puppet
  , PuppetWeb
  , Room
  , Sayable
  , UtilLib
  , Wechaty
  , log // for convenionce use npmlog with environment variable LEVEL
  , StateMonitor
  , Brolog
}
```

runDocker.sh
```
cp facade.docker.ts facade.ts
```

mysource.ts
```javascript
import { Config }           from './facade'
import { Io }               from './facade'
import { Message }          from './facade'
import { StateMonitor }     from './facade'
import { Wechaty }          from './facade'
import { Brolog }           from './facade'
... ... some other import
```


 here i use  following code to isolate FriendRequest problem in my source:
```javascript
import { FriendRequest }  from 'wechaty/src/puppet-web/friend-request'
```

**Comment by zixia** [comment]
@JasLin agree. this is definitely a good idea for now.

**Comment by JasLin** [comment]
@zixia verified! it works.

**Comment by JasLin** [comment]
@zixia do you have any iead how to fix the path reference problem while run wechaty with docker or develop with wechaty  source code at the same time. 

since typescript do not allow to make ```import path ``` dynamic changing at runtime . 

ref: http://stackoverflow.com/questions/18118222/dynamically-import-module-in-typescript

**Comment by zixia** [comment]
Ah, it should be easy, sorry for forgot to document that. 

I think the best practice for this is to install wechaty via NPM(or npm link) in your develop environment, and then you can use `import Wechaty from 'wechaty'` .

For example:

```shell
$ mkdir wechaty-bot
$ cd wechaty-bot
$ npm init -y
$ npm install wechaty # or `npm link wechaty` if you had linked source code to npm global
$ cat > mybot.ts
import { Wechaty } from 'wechaty'
console.log(Wechaty.instance().version())
^D

$ ts-node mybot.ts # this should work
$ wechaty mybot.ts # this should work too.
```

**Comment by JasLin** [comment]
ah, yes ,use npm install wechaty is easy. 
while i am developing under wechaty source like this:

```javascript
cd wechaty
git clone  some@somehost:robots 
```
all the codes are in ```robots```

because i'm using some low level api , such as ``` wechaty/src/puppet-web/friend-request``` at before.

so i should move all the code to a project with npm install ```wechaty```,  and check out which low level was used , and then make pull request to you:)

**Comment by zixia** [comment]
yes, please. 

it seems we are on our way moving from alpha to beta stage since now. :)
